api/envoy/config/core/v3/base.proto

```diff
@@ -303,12 +303,31 @@ message RuntimeFeatureFlag {
   string runtime_key = 2 [(validate.rules).string = {min_len: 1}];
 }
 
+// Please use :ref:`KeyValuePair <envoy_api_msg_config.core.v3.KeyValuePair>` instead.
+// [#not-implemented-hide:]
 message KeyValue {
+  // The key of the key/value pair.
+  string key = 1 [
+    deprecated = true,
+    (validate.rules).string = {min_len: 1 max_bytes: 16384},
+    (envoy.annotations.deprecated_at_minor_version) = "3.0"
+  ];
+
+  // The value of the key/value pair.
+  //
+  // The ``bytes`` type is used. This means if JSON or YAML is used to to represent the
+  // configuration, the value must be base64 encoded. This is unfriendly for users in most
+  // use scenarios of this message.
+  //
+  bytes value = 2 [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
+}
+
+message KeyValuePair {
   // The key of the key/value pair.
   string key = 1 [(validate.rules).string = {min_len: 1 max_bytes: 16384}];
 
   // The value of the key/value pair.
-  bytes value = 2;
+  google.protobuf.Value value = 2;
```

Comment:
> I'm assuming the bytes type above was used to allow utf-8 characters. I don't think that the protobuf.Value supports that (but maybe I'm wrong).

The most scenarios that the message design for is using normal utf-8. The `bytes` is used because we didn't notice it's side effect when we designed it.
Keep the `bytes value` will bring lots of confusion because it cannot be used like most our `value` field in HeaderValue, etc.

If one day the binary's requirement (as a corner case) is raised, we can support a special substitution format to support bytes here by `%BASE64_DECODE(xxxxx)%` in the future.

Reply:
> Yeah, I guess if there are no other uses, it is probably ok to start this way. I do wonder why this type isn't used though...
> 
> If it were used I would definitely vote in favor of adding another field to the original message.

Yeah, if it's used, this would be complete different way. And when I created this type, it also supprised me and @htuch why we don't have a type like this. (although we still made (`bytes` value) a mistake when we create this message).